src/sm/launcher/launcher.cpp

```diff
@@ -962,4 +1057,208 @@ Error Launcher::HandleOfflineTTLs()
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error Launcher::GetInstanceEnvVars(
+    const InstanceIdent& instanceIdent, Array<StaticString<cEnvVarNameLen>>& envVars) const
+{
+    for (const auto& currentEnvVar : mCurrentEnvVars) {
+        if (currentEnvVar.mFilter.Match(instanceIdent)) {
+            for (const auto& envVar : currentEnvVar.mVariables) {
+                if (auto err = envVars.PushBack(envVar.mName); !err.IsNone()) {
+                    return AOS_ERROR_WRAP(err);
+                }
+            }
+        }
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error Launcher::GetEnvChangedInstances(Array<InstanceData>& instances) const
+{
+    for (auto& instance : mCurrentInstances) {
+        auto envVars = MakeUnique<EnvVarsArray>(&mAllocator);
+
+        if (auto err = GetInstanceEnvVars(instance.Info().mInstanceIdent, *envVars); !err.IsNone()) {
+            LOG_ERR() << "Can't get instance env vars: instanceID=" << instance.InstanceID() << ", err=" << err;
+
+            continue;
+        }
+
+        if (*envVars == instance.GetOverrideEnvVars()) {
+            continue;
+        }
+
+        if (auto err = instances.EmplaceBack(instance.Info(), instance.InstanceID()); !err.IsNone()) {
+            return AOS_ERROR_WRAP(err);
+        }
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error Launcher::SendEnvChangedInstancesStatus(const Array<InstanceData>& instances)
+{
+    auto status = MakeUnique<InstanceStatusStaticArray>(&mAllocator);
+
+    for (const auto& instanceData : instances) {
+        auto instance = GetInstance(instanceData.mInstanceID);
+        if (instance == mCurrentInstances.end()) {
+            LOG_ERR() << "Can't get instance: instanceID=" << instanceData.mInstanceID
+                      << ", err=" << Error(ErrorEnum::eNotFound);
+
+            continue;
+        }
+
+        if (instance->RunError().IsNone()) {
+            LOG_DBG() << "Run instance status: instanceID=" << *instance
+                      << ", serviceVersion=" << instance->GetServiceVersion() << ", runState=" << instance->RunState();
+        } else {
+            LOG_ERR() << "Run instance status: instanceID=" << *instance
+                      << ", serviceVersion=" << instance->GetServiceVersion() << ", runState=" << instance->RunState()
+                      << ", err=" << instance->RunError();
+        }
+
+        if (auto err = status->PushBack({instance->Info().mInstanceIdent, instance->GetServiceVersion(),
+                instance->RunState(), instance->RunError()});
+            !err.IsNone()) {
+            return AOS_ERROR_WRAP(err);
+        }
+    }
+
+    LOG_DBG() << "Send update status";
+
+    if (auto err = mStatusReceiver->InstancesUpdateStatus(*status); !err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error Launcher::UpdateInstancesEnvVars()
+{
+    auto restartInstances = SharedPtr<Array<InstanceData>>(&mAllocator, new (&mAllocator) InstanceDataStaticArray());
```

Comment:
Can we use make shared?
```suggestion
    auto restartInstances = MakeShared<InstanceDataStaticArray>(mAllocator);
```